dj/construction/build.py

```diff
@@ -0,0 +1,275 @@
+            )
+            alias = amenable_name(node.name)
+            node_select = node_query.select
+            node_ast = ast.Alias(ast.Name(alias), child=node_select)
+            for tbl in tbls:
+                select.replace(tbl, node_ast)
+        else:
+            node_table = [
+                table for table in node.tables if table.database.id == database.id
+            ][0]
+            node_ast = ast.Table(
+                ast.Name(node_table.table),
+                namespace=ast.Namespace(
+                    [
+                        ast.Name(s)
+                        for s in (node_table.catalog, node_table.schema_)
+                        if s
+                    ],
+                ),
+            )
+
+        for tbl in tbls:
+            select.replace(tbl, node_ast)
+
+    return select
+
+
+def build_query(  # pylint: disable=too-many-arguments
+    session: Session,
+    query: ast.Query,
+    build_plan: BuildPlan,
+    build_plan_depth: int,
+    database: Database,
+    dialect: Optional[str] = None,
+) -> ast.Query:
+    """transforms a query ast by replacing dj node references with their asts"""
+    select = query.to_select()
+    build_select(session, select, build_plan, build_plan_depth, database, dialect)
+    for i, exp in enumerate(select.projection):
+        if not isinstance(exp, ast.Named):
+            name = f"_col{i}"
+            aliased = ast.Alias(ast.Name(name), child=exp)
+            # only replace those that are identical in memory
+            select.replace(exp, aliased, lambda a, b: id(a) == id(b))
+    return query
+
+
+def add_filters_aggs(
+    query: ast.Query,
+    dialect: Optional[str] = None,
+    filters: Optional[List[str]] = None,
+    aggs: Optional[List[str]] = None,
+) -> ast.Query:
+    """add filters and aggs to a query ast"""
+    if filters:
+        filter_asts = (  # pylint: disable=consider-using-ternary
+            query.select.where and [query.select.where] or []
+        )
+
+        for filter_ in filters:
+            filter_asts.append(
+                parse(f"select * where {filter_}", dialect).select.where,  # type:ignore
+            )
+        query.select.where = reduce(
+            lambda left, right: ast.BinaryOp(
+                ast.BinaryOpKind.And,
+                left,
+                right,
+            ),
+            filter_asts,
+        )
+    if aggs:
+        for agg in aggs:
+            query.select.group_by += parse(
+                f"select * group by {agg}",
+                dialect,
+            ).select.group_by  # type:ignore
+
+    return query
+
+
+async def build_node(  # pylint: disable=too-many-arguments
+    session: Session,
+    node: Node,
+    dialect: Optional[str] = None,
+    database_id: Optional[int] = None,
+    filters: Optional[List[str]] = None,
+    aggs: Optional[List[str]] = None,
+) -> Tuple[ast.Query, Database]:
+    """transforms a query ast by replacing dj node references with their asts
```

Comment:
What about "Determines the optimal database to run the query in and builds the query AST appropriately"?

Reply:
done